### ISSUE TITLE
GraphJSON v1 Deserializer

### ISSUE THREAD

**Issue by TimothyDespair** [Revmaker/gremlex#76]
Just getting started with Elixir so I may be missing something obvious here, but I've followed the steps in the readme using Titan running a gremlin server. It connected fine and I ran a query on the empty database that returned just fine:

```
iex(3)> Graph.g() |> Graph.v() |> Client.query
{:ok, []}
```

After jumping into Titan's gremlin shell and adding a node like such:

```
gremlin> :> graph.addVertex("test", "test")
==>v[4144]
```

And then running the query again, I get an error like such:

```
iex(4)> Graph.g() |> Graph.v() |> Client.query
[error] Task #PID<0.2338.0> started from #PID<0.2163.0> terminating
** (CaseClauseError) no case clause matching: [%{"id" => 4144, "label" => "vertex", "properties" => %{"test" => [%{"id" => "sm-374-sl", "value" => "test"}]}, "type" => "vertex"}]
    (gremlex) lib/gremlex/deserializer.ex:8: Gremlex.Deserializer.deserialize/1
    (gremlex) lib/gremlex/client.ex:73: Gremlex.Client.recv/2
    (elixir) lib/task/supervised.ex:89: Task.Supervised.do_apply/2
    (elixir) lib/task/supervised.ex:38: Task.Supervised.reply/5
    (stdlib) proc_lib.erl:249: :proc_lib.init_p_do_apply/3
Function: #Function<0.105175084/0 in Gremlex.Client.handle_call/3>
    Args: []
** (exit) exited in: GenServer.call(#PID<0.2163.0>, {:query, "{\"requestId\":\"a104bae6-c9be-48e0-b035-1e3b17891681\",\"processor\":\"\",\"op\":\"eval\",\"args\":{\"language\":\"gremlin-groovy\",\"gremlin\":\"g.V()\"}}"}, 5000)
    ** (EXIT) an exception was raised:
        ** (CaseClauseError) no case clause matching: [%{"id" => 4144, "label" => "vertex", "properties" => %{"test" => [%{"id" => "sm-374-sl", "value" => "test"}]}, "type" => "vertex"}]
            (gremlex) lib/gremlex/deserializer.ex:8: Gremlex.Deserializer.deserialize/1
            (gremlex) lib/gremlex/client.ex:73: Gremlex.Client.recv/2
            (elixir) lib/task/supervised.ex:89: Task.Supervised.do_apply/2
            (elixir) lib/task/supervised.ex:38: Task.Supervised.reply/5
            (stdlib) proc_lib.erl:249: :proc_lib.init_p_do_apply/3
    (elixir) lib/gen_server.ex:924: GenServer.call/3
    (poolboy) src/poolboy.erl:76: :poolboy.transaction/3
```

I've had a read of the deserialiser code. And it seems to be expecting a very different structure for `result["data"]` than what it is being passed. I'm not sure if the "@type" etc. and "g:List" etc. are special guards I'm not familiar with or the actual keys that `deserialise` is expecting. But even beyond that it seems to expect the keys `value` and `type` to be in the `result["data"]` map while `value` is currently inside the `result["data"]["properties"]["test"]` map.

Honestly the structure of the vertex is confusing as hell, which makes me think this is partly weirdness with Titan's vertex structure. But is it worth implementing a more structure agnostic deserialiser? I'm happy to take a swing at it and submit a PR. If I can wrap my head around how "g:____" and "@____" work in the guards I'm sure I can get it working easily enough, but google isn't yielding any answers so any explanation of how that works is greatly appreciated!

**Comment by kevmojay** [comment]
I haven't used titan yet but it looks like they return a different structure like you noticed.

You aren't missing anything obvious. 

Here is a sample response from the Gremlin server for what I typically use

`%{
  "requestId" => "2d32e2d1-60d8-4bd0-b7e4-299c5467c73a",
  "result" => %{
    "data" => %{
      "@type" => "g:List",
      "@value" => [
        %{
          "@type" => "g:Vertex",
          "@value" => %{
            "id" => %{"@type" => "g:Int64", "@value" => 16},
            "label" => "Intent",
            "properties" => %{
              "name" => [
                %{
                  "@type" => "g:VertexProperty",
                  "@value" => %{
                    "id" => %{"@type" => "g:Int64", "@value" => 17},
                    "label" => "name",
                    "value" => "request.details"
                  }
                }
              ]
            }
          }
        },`

This is more or less the structure deserialize is expecting.

So as you can see the "g:" and "@" aren't anything special, just simply the key in the map. 

i haven't gotten the chance to look at what titan responds with but I am guessing it won't be too difficult to add support.

If you want to take a swing at a PR, just inspect the structure by inspecting the result [here](https://github.com/Revmaker/gremlex/blob/master/lib/gremlex/client.ex#L78)

And [then in this case statement](https://github.com/Revmaker/gremlex/blob/master/lib/gremlex/deserializer.ex#L8) add a new pattern against what you see in `result["data"]` assuming that is what Titan returns.

I think from there, you will be able to have a good idea on what to do next.

I also don't mind working towards a PR to support Titan.

**Comment by TimothyDespair** [comment]
Awesome! I'm going to start playing around with this now.

**Comment by barakyo** [comment]
@TimothyDespair I was going to jump into this to try and help out, but the first thing I noticed when I googled "TitanDB" was a couple of links like:

* [Is TitanDB Dead?](https://github.com/thinkaurelius/titan/issues/1360)
* [JanusGraph Picks Up Where TitanDB Left Off - Datanami](https://www.datanami.com/2017/01/13/janusgraph-picks-titandb-left-off/)

Do you believe its worth supporting TitanDB or should we consider an alternative like JanusGraph?

**Comment by barakyo** [comment]
@TimothyDespair just to follow up, I did some superficial testing and it looks like Gremlex works with JanusGraph (at least to some extent)

```
iex(7)> Graph.g() |> Graph.add_v("foo") |> Graph.property("name", "kevmojay") |> Client.query
{:ok,
 [
   %Gremlex.Vertex{
     id: "7cf39161-5768-43e7-ba1e-828cce754313",
     label: "foo",
     properties: %{name: ["kevmojay"]}
   }
 ]}
iex(8)> Graph.g() |> Graph.v() |> Client.query
{:ok,
 [
   %Gremlex.Vertex{
     id: "e7dc5cba-9505-4192-b95e-0d42328cf3b7",
     label: "foo",
     properties: %{name: ["barakyo"]}
   },
   %Gremlex.Vertex{
     id: "7cf39161-5768-43e7-ba1e-828cce754313",
     label: "foo",
     properties: %{name: ["kevmojay"]}
   },
   %Gremlex.Vertex{
     id: "4e2b33a8-6273-46fe-a75d-420bdeaa2fce",
     label: "foo",
     properties: %{}
   }
 ]}
```

**Comment by TimothyDespair** [comment]
Ah! Crap, I do remember reading that when I was doing my research, I think I just got my wires crossed remembering which one had been abandoned. Should have spotted the last commit being 3 years ago on Titan.

I’ll fire up a Janus database today and see how it goes.

**Comment by barakyo** [comment]
Sounds good! Good luck! 🙏🏽 Keep us posted!

**Comment by TimothyDespair** [comment]
Ah! I'd been running 0.2.2 which was incorrectly marked as the latest release on their releases page. I seem to be getting the same behaviour as you now. Sorry about the run-around!

**Comment by barakyo** [comment]
Awesome! Glad to hear that! I linked the specific release for that reason. I had to search through the release page to find the latest one as well. Let us know how JanuaGraph goes!

**Comment by TimothyDespair** [comment]
To be at least a little useful I've found the culprit for it all: [GraphSON v2](https://github.com/apache/tinkerpop/pull/351)

Titan and the earlier Janus release were still using GraphSON v1, newer releases are using GraphSON v2.

I don't know how much backwards compatibility matters, but writing a deserialiser for v1 was pretty trivial (only 3 or 4 function overloads, although now the distinction is clearer I'd clean it up a little). Because it's not typed there's not much to do except strip ids out of properties. It was a little messy to do by overloading the deserialize function, but it'd be easy enough to detect and switch at the `result["data"]` case on [line 8](https://github.com/Revmaker/gremlex/blob/c674629a93c3c37171239ffd8a075970e8c6410b/lib/gremlex/deserializer.ex#L8) and call a `deserialize_legacy` or something similar. If I've got the time to tinker over the next couple of days I'll come back to it, but I've got to move on to some other stuff now I've got this up and running.

Updated the title _again_ to reflect the actual problem so it's quicker to identify for anyone in the future who comes across it.

I'll keep you posted on how Janus goes!
